src/PactNet.Native/NativePactMessageBuilder.cs

```diff
@@ -0,0 +1,168 @@
+        void IPactMessageBuilderV3.Build()
+            => Build();
+
+        /// <inheritdoc cref="IPactMessageBuilderV3"/>
+        void IPactMessageBuilderV3.Verify<T>(Action<T> handler)
+            => Verify(handler);
+
+        #endregion
+
+        #region Internal Methods
+
+        /// <summary>
+        /// Add a new message to the pact
+        /// </summary>
+        /// <param name="description">Message description</param>
+        /// <returns>Fluent builder</returns>
+        internal IPactMessageBuilderV3 ExpectsToReceive(string description)
+        {
+            server.MessageExpectsToReceive(message, description);
+
+            return this;
+        }
+
+        /// <summary>
+        /// Add a provider state
+        /// </summary>
+        /// <param name="providerState">Provider state description</param>
+        /// <returns>Fluent builder</returns>
+        internal IPactMessageBuilderV3 Given(string providerState)
+        {
+            server.MessageGiven(message, providerState);
+
+            return this;
+        }
+
+        /// <summary>
+        /// Add a provider state with one or more parameters
+        /// </summary>
+        /// <param name="providerState">Provider state description</param>
+        /// <param name="parameters">Provider state parameters</param>
+        /// <returns>Fluent builder</returns>
+        internal IPactMessageBuilderV3 Given(string providerState, IDictionary<string, string> parameters)
+        {
+            foreach (var param in parameters)
+            {
+                server.MessageGivenWithParam(message, providerState, param.Key, param.Value);
+            }
+
+            return this;
+        }
+
+        /// <summary>
+        /// Set the metadata
+        /// </summary>
+        /// <param name="key">key of the metadata</param>
+        /// <param name="value">value of the metadata</param>
+        /// <returns>Fluent builder</returns>
+        internal IPactMessageBuilderV3 WithMetadata(string key, string value)
+        {
+            server.MessageWithMetadata(message, key, value);
+
+            return this;
+        }
+
+        /// <summary>
+        /// Set the content
+        /// </summary>
+        /// <param name="content">Dynamic content</param>
+        /// <returns>Fluent builder</returns>
+        internal IPactMessageBuilderV3 WithContent(dynamic content)
+        {
+            server.MessageWithContents(message, "application/json", JsonConvert.SerializeObject(content), 100);
+
+            return this;
+        }
+
+        /// <summary>
+        /// Build the pact file
+        /// </summary>
+        internal void Build()
+        {
+            server.WriteMessagePactFile(pact, config.PactDir, true);
+        }
+
+        /// <summary>
+        /// Verify a message is read and handled correctly
+        /// </summary>
+        /// <param name="handler">The method using the message</param>
+        internal void Verify<T>(Action<T> handler)
```

Comment:
Probabably needs an async version also to match the request/response verification.